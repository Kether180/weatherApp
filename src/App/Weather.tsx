import React from "react";

interface WeatherInfo {
  city: string | null;
  temp: number | null;
  icon: string | null;
  description: string | null;
  humidity: number | null;
  windSpeed: number | null;
}

interface LocationInfo {
  name: string;
  country: string;
  state: string;
}

interface WeatherProps {
  city: string | null;
  weatherData?: WeatherInfo;
}

const Weather = ({ city, weatherData }: WeatherProps) => {
  const [info, setInfo] = React.useState<WeatherInfo>({
    city: null,
    temp: null,
    icon: null,
    description: null,
    humidity: null,
    windSpeed: null,
  });
  const [locationInfo, setLocationInfo] = React.useState<LocationInfo | null>(
    null
  );
  const [error, setError] = React.useState<string | null>(null);
  const [isPending, setIsPending] = React.useState(false);

  React.useEffect(() => {
    if (weatherData) {
      setInfo(weatherData);
      return;
    }

    if (city !== null || isPending === false) {
      setIsPending(true);

      fetch(
        `https://api.openweathermap.org/geo/1.0/direct?q=${city}&limit=1&appid=4c4f0b1876954338598a7be96c66527b`
      )
        .then((res) => res.json())
        .then((data) => {
          console.log(data);

          if (data.length > 0) {
            setLocationInfo({
              name: data[0].name,
              country: data[0].country,
              state: data[0].state,
            });

            fetch(
              `https://api.openweathermap.org/data/2.5/weather?lat=${data[0].lat}&lon=${data[0].lon}&units=metric&appid=4c4f0b1876954338598a7be96c66527b`
            )
              .then((res) => res.json())
              .then((data) => {
                console.log(data);

                setInfo({
                  city: data.name,
                  temp: data.main.temp,
                  icon: `https://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`,
                  description: data.weather[0].description,
                  humidity: data.main.humidity,
                  windSpeed: data.wind.speed,
                });

               setIsPending(false);
              })
              .catch((error) => {
                console.log(error);
                setError("Error fetching weather data");
              });
          }
        })
        .catch((error) => {
          console.log(error);
          setError("Error fetching location data");
        });
    }
  }, [city, weatherData, isPending]);

  return (
    <>
      {error && <div>{error}</div>}
      {!error && (
        <>
          {locationInfo && (
            <div>
              <h2>
                {locationInfo.name}, {locationInfo.state},{" "}
                {locationInfo.country}
              </h2>
            </div>
          )}
          <div>
            {info.description} 
            </div>
          <img 
          src={info.icon ?? ""} alt="" />
          <div> 
             {info.temp && ~~info.temp} Celsius
            </div>
          <div>
             Humidity: {info.humidity}
          %</div>
          <div> <p>
          Wind Speed: {info.windSpeed} 
             km/h</p>
            </div>
        </>
      )}
    </>
  );
};

export { Weather };
