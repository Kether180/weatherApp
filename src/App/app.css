:root {
  font-size: 13px;
  font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
  --dark: #0f0707;
  --light: #e6e7e0;
  --teal: #f5f5aa;
}

/* Define your color palette */
:root[data-theme="light"] {
  --background-color: var(--light);
  --text-color: var(--dark);
}

:root[data-theme="dark"] {
  --background-color: var(--dark);
  --text-color: var(--light);
}

html {
  height: 100%;
  width: 100%;
  display: grid;
  place-items: center;
  margin: 20px;
  padding: 100px;
  background-color: var(--background-color);
  color: var(--text-color);
}

body {
  min-height: 100vh;
  background-color: var(--background-color);
  color: var(--text-color);
  place-items: center;
  padding: 30px;
}

* {
  margin: 0;
  padding: 0;
  border: none;
  font-size: 1.5rem;
  text-transform: none;
  text-decoration: none;
  box-sizing: border-box;
  margin: 10px;

}

div {

  border-radius: 0rem;
  display: grid;
  grid-template: repeat(3, min-content) / repeat(2, 25rem);
  grid-template-areas:
    "input submit"
    "city temp"
    "icon icon";
  gap: 1rem;
  text-align: center;
  padding: -9rem;
  margin: 0rem;


}

div,
p {
  font-size: 2rem;
  font-weight: 390;
  text-align: center;
  padding: 2rem;
  margin: -2.5rem;


}


input,
button {
  height: 3rem;
  border-radius: 1rem;
  text-align: right;
}



label,
#light-mode-switch {
  height: 2rem;
  border-radius: 3rem;
  text-align: right;
}

input {
  grid-area: input;
  padding-inline: 1rem;
  text-align: center;
  border: 2px solid var(--dark);
}

button {
  grid-area: submit;
  cursor: pointer;
  border: 2px solid var(--teal);
  background-color: var(--dark);
  color: var(--teal);
  transition: transform 0.15s ease-in-out, border-color 0.15s ease-in-out, color 0.15s ease-in-out;
  text-align: center;
}

button:hover,
button:focus {
  background-color: var(--light);
  color: var(--dark);
  border-color: var(--light);
  transform: scale(1.05);
  text-align: center;
}

h1 {
  grid-area: city;
  text-align: center;
  font-size: 3rem;
}


p {
  grid-area: temp;
}

img {
  grid-area: icon;
  justify-self: center;
  margin: -3rem;
  height: 10rem;
}

/* Responsive styles  , mobile */

@media only screen and (max-width: 767px) {
  :root {
    font-size: 42.5%;
    /* 1rem = 10px */
    font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
    --dark: #0f0707;
    --light: #e6e7e0;
    --teal: #f5f5aa;
  }

  :root[data-theme="light"] {
    --background-color: var(--light);
    --text-color: var(--dark);
  }

  :root[data-theme="dark"] {
    --background-color: var(--dark);
    --text-color: var(--light);
  }

  html {
    height: 100%;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: var(--background-color);
    color: var(--text-color);
  }

  body {
    min-height: 100vh;
    background-color: var(--background-color);
    color: var(--text-color);
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 0;
    margin: 0;
  }

  input,
  button,
  label,
  #light-mode-switch {
    margin: 0;
    padding: 0;
    border: none;
    font-size: 1.5rem;
    text-transform: none;
    text-decoration: none;
    box-sizing: border-box;
  }

  input {
    padding: 0 1rem;
    text-align: center;
    border: 2px solid var(--dark);
    height: 3rem;
    border-radius: 1rem;
  }

  button {
    cursor: pointer;
    border: 2px solid var(--teal);
    background-color: var(--dark);
    color: var(--teal);
    transition: transform 0.15s ease-in-out, border-color 0.15s ease-in-out, color 0.15s ease-in-out;
    text-align: center;
    height: 3rem;
    border-radius: 1rem;
  }

  button:hover,
  button:focus {
    background-color: var(--light);
    color: var(--dark);
    border-color: var(--light);
    transform: scale(1.05);
  }

  div {
    grid-template: repeat(3, min-content) / 1fr;
    grid-template-areas:
      "input"
      "submit"
      "city"
      "temp"
      "icon";
    padding: 2rem;
  }


  div,
  p {
    font-size: 2.2rem;
    line-height: 1.5;
    text-align: center;
    padding: 2rem;
    margin: -2rem;
  }

  label,
  #light-mode-switch {
    height: 2rem;
    border-radius: 3rem;
    text-align: right;
  }

  img {
    margin: -1.5rem;
    height: 7.5rem;
  }

  h1 {
    grid-area: city;
    text-align: center;
    font-size: 10rem;
  }


  @media only screen and (min-width: 1025px) and (max-width: 1280px) {


    div {
      grid-template: repeat(3, min-content) / 1fr;
      grid-template-areas:
        "input"
        "submit"
        "city"
        "temp"
        "icon";
      padding: 2rem;
    }


    div,
    p {
      font-size: 2rem;
      font-weight: 390;
      text-align: center;
      padding: 2rem;
      margin: -2rem;


    }


    input,
    button {
      height: 3rem;
      border-radius: 1rem;
      text-align: right;
    }



    label,
    #light-mode-switch {
      height: 2rem;
      border-radius: 3rem;
      text-align: right;
    }

    h1 {
      grid-area: city;
      text-align: center;
      font-size: 3rem;
    }


  }

}