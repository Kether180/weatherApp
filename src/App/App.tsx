import React, { ChangeEvent, useState, FC } from "react";

import { Weather } from "./Weather";

import "./app.css";

import { ThemeProvider } from "../ThemeProvider";

export const App: FC = () => {
  const [input, setInput] = useState("");

  const [city, setCity] = useState("");

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setInput(event.target.value);
  };

  const handleShowWeatherClick = () => {
    const trimmedInput = input.trim();
    if (trimmedInput !== "") {
      setCity(input);
    }
  };

  return (
    <ThemeProvider>
      <>
        <input
          role="search"
          type="text"
          value={input}
          onChange={handleInputChange}
          placeholder="Search for a city"
        />
        <button onClick={handleShowWeatherClick}>Show Weather</button>
        <Weather city={city} />
      </>
    </ThemeProvider>
  );
};
