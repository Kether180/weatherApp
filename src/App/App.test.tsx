import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import { App } from "./App"

describe( "App", () => {
  it( "Renders Component According To Snapshot", () => {
    const { container } = render( <App /> )
    
    expect( container ).toMatchSnapshot() } )

  it( "Checks Input Functionality", () => {
    render( <App /> )

    userEvent.type( screen.getByRole( "search" ), "Copenhagen" )
    expect( screen.getByRole( "search" ) ).toHaveValue( "Copenhagen" ) } ) } )