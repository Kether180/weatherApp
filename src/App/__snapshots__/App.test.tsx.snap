// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders Component According To Snapshot 1`] = `
<div>
  <button>
    Toggle Theme
  </button>
  <div>
    <label
      for="light-mode-switch"
    />
    <input
      checked=""
      id="light-mode-switch"
      type="checkbox"
    />
    <div>
      (I/O)
    </div>
  </div>
  <input
    placeholder="Search for a city"
    role="search"
    type="text"
    value=""
  />
  <button>
    Show Weather
  </button>
  <div />
  <img
    src=""
  />
  <div>
     Celsius
  </div>
  <div>
    Humidity: 
    %
  </div>
  <div>
    Wind Speed: 
    km/h
  </div>
</div>
`;
